coresdk/src/backend/gpio_driver.cpp

```diff
@@ -104,5 +112,153 @@ namespace splashkit_lib
+            set_pud_cmd.param1 = pin;
+            set_pud_cmd.param2 = pud;
+
+            sk_gpio_send_cmd(pi, set_pud_cmd);
+        }
+
+        int sk_remote_gpio_read(connection pi, int pin)
+        {
+            sk_pigpio_cmd_t read_cmd;
+            read_cmd.cmd_code = GPIO_CMD_READ;
+            read_cmd.param1 = pin;
+
+            return sk_gpio_send_cmd(pi, read_cmd);
+        }
+
+        void sk_remote_gpio_write(connection pi, int pin, int value)
+        {
+            sk_pigpio_cmd_t write_cmd;
+            write_cmd.cmd_code = GPIO_CMD_WRITE;
+            write_cmd.param1 = pin;
+            write_cmd.param2 = value;
+
+            sk_gpio_send_cmd(pi, write_cmd);
+        }
+
+        void sk_remote_set_pwm_range(connection pi, int pin, int range)
+        {
+            sk_pigpio_cmd_t set_range_cmd;
+            set_range_cmd.cmd_code = GPIO_CMD_SET_PWM_RANGE;
+            set_range_cmd.param1 = pin;
+            set_range_cmd.param2 = range;
+
+            sk_gpio_send_cmd(pi, set_range_cmd);
+        }
+
+        void sk_remote_set_pwm_frequency(connection pi, int pin, int frequency)
+        {
+            sk_pigpio_cmd_t set_freq_cmd;
+            set_freq_cmd.cmd_code = GPIO_CMD_SET_PWM_FREQ;
+            set_freq_cmd.param1 = pin;
+            set_freq_cmd.param2 = frequency;
+
+            sk_gpio_send_cmd(pi, set_freq_cmd);
+        }
+
+        void sk_remote_set_pwm_dutycycle(connection pi, int pin, int dutycycle)
+        {
+            sk_pigpio_cmd_t set_dutycycle_cmd;
+            set_dutycycle_cmd.cmd_code = GPIO_CMD_SET_PWM_DUTYCYCLE;
+            set_dutycycle_cmd.param1 = pin;
+            set_dutycycle_cmd.param2 = dutycycle;
+
+            sk_gpio_send_cmd(pi, set_dutycycle_cmd);
+        }
+
+        void sk_remote_clear_bank_1(connection pi)
+        {
+            sk_pigpio_cmd_t clear_bank_cmd;
+            clear_bank_cmd.cmd_code = GPIO_CMD_CLEAR_BANK_1;
+            clear_bank_cmd.param1 = 0x0FFFFFFC;
+
+            sk_gpio_send_cmd(pi, clear_bank_cmd);
+
+        }
+
+        bool sk_remote_gpio_cleanup(connection pi)
+        {
+            if(!is_connection_open(pi))
+            {
+                LOG(ERROR) << "Remote GPIO: Connection not open.";
+                return false;
+            }
+			cout << "Cleaning Pins on Remote Pi Named: " << pi->name << endl;
+            sk_remote_clear_bank_1(pi);
+            return close_connection(pi);
+        }
+
+        int sk_gpio_send_cmd(connection pi, sk_pigpio_cmd_t &cmd)
+        {
+            if(!is_connection_open(pi))
+            {
+                LOG(ERROR) << "Remote GPIO: Connection not open.";
+                return -1;
+            }
+
+            if(pi->protocol == TCP)
+            {
+                int num_send_bytes = sizeof(cmd);
+
+                char buffer[num_send_bytes];
+                memcpy(buffer, &cmd, num_send_bytes);
+
+                if(sk_send_bytes(&pi->socket, buffer, num_send_bytes)) 
```

Comment:
I'm wondering why the command is copied to the buffer? E.g. would it be possible to do

```suggestion
                char buffer[num_send_bytes];
                
                if(sk_send_bytes(&pi->socket, cmd, num_send_bytes)) 
```

Reply:
Nah this doesn't work, what you can do:

`if(sk_send_bytes(&pi->socket, reinterpret_cast<char*>(&cmd), num_send_bytes))`

and then

```
sk_pigpio_cmd_t resp;
int num_bytes_recv = sk_read_bytes(&pi->socket, reinterpret_cast<char*>(&resp), num_send_bytes);
if(num_bytes_recv == num_send_bytes)
{
    return resp.result;
}
```

IMO this is not ideal however, as how the data is interpreted is dependent on the system. Check out the example here: https://en.cppreference.com/w/cpp/language/reinterpret_cast

```
int i = 7;
 
...
 // type aliasing through pointer
char* p2 = reinterpret_cast<char*>(&i);
    std::cout << (p2[0] == '\x7' ? "This system is little-endian\n"
 : "This system is big-endian\n");
```

So, whether that works correctly is dependent on what we run it on.

Reply:
Sh I see because the response of the command is in the command struct. I need to look more into pigio but would splitting the command and the response into seperate structs be possible?

Either way I don't think this needs to block the PR being merged.

Reply:
As far as I can tell, no, but I'll have to dive into the code itself to confirm this.